### DISCUSSION TITLE
Spack Package for DIALS

### DISCUSSION THREAD

**Discussion by RobertRosca** [dials/dials#1589] [Ideas]
Hey, I made this PR (https://github.com/dials/dials/pull/1534) a while ago to fix a small issue with the `setup.py` file and briefly mentioned that we encountered that issue while creating Spack packages for a number of tools used at European XFEL.

For some context, here's a brief description of what Spack is:

> Spack is a package management tool designed to support multiple versions and configurations of software on a wide variety of platforms and environments. It was designed for large supercomputing centers, where many users and application teams share common installations of software on clusters with exotic architectures, using libraries that do not have a standard ABI. Spack is non-destructive: installing a new version does not break existing installations, so many configurations can coexist on the same system.

You can find more details on the Spack repo (https://github.com/spack/spack/) or on their docs pages (https://spack.readthedocs.io/en/latest/).

Basically, Spack lets you define the build process for some software as a Python class, a lot of [build systems are supported](https://spack.readthedocs.io/en/latest/build_systems.html), with Python being one of them. Spack has thousands of packages defined already, including hundreds of Python packages.

Creating a 'standard' setuptools-based Python package in Spack is pretty simple, here are some examples:
- [h5glance](https://github.com/spack/spack/blob/develop/var/spack/repos/builtin/packages/py-h5glance/package.py)
- [tomopy](https://github.com/spack/spack/blob/develop/var/spack/repos/builtin/packages/py-tomopy/package.py)

It's also possible to make Spack packages for more complex Python packages as well:
- [numpy](https://github.com/spack/spack/blob/develop/var/spack/repos/builtin/packages/py-numpy/package.py)
- [scipy](https://github.com/spack/spack/blob/develop/var/spack/repos/builtin/packages/py-scipy/package.py)
- [matplotlib](https://github.com/spack/spack/blob/develop/var/spack/repos/builtin/packages/py-matplotlib/package.py)
- [pytorch](https://github.com/spack/spack/blob/develop/var/spack/repos/builtin/packages/py-torch/package.py)

With pytorch being probably one of the more complex Spack Python packages, as it's a CUDA package as well as a python one, and has a lot of compilation/build environment options.

I've been working with @julianhoersch (student hired to help with this project) on creating Spack packages for our use at European XFEL, and we'd like to make a Spack package for DIALS as well, but the setup/bootstrapping process is pretty complex and I'm not sure if it would be possible for us to specify all of the build options and steps ourselves.

Would any of the DIALS maintainers be interested in helping us make and test a Spack package for DIALS? If this seems useful then I'd be happy to set up a meeting so we could discuss this more.

Thanks,

Robert

**Comment by Anthchirp** [comment]
I believe, as is often the case with these things, your main problem will be to get the DIALS dependencies working first. In particular, you'll need cctbx from somewhere. We won't be able to help with that.

Basically, once you have cctbx and dxtbx (and all their dependencies - cbflib comes to mind) sorted, DIALS itself should be easy.
